### ISSUE TITLE
Merge Johnny the GUI into the jumbo tree & build

### ISSUE THREAD

**Issue by solardiz** [magnumripper/JohnTheRipper#2673]
I think Johnny the GUI will receive more attention from both (prospective) maintainers/developers and users if we merge it into jumbo. Have it as a configure option (maybe enabled by default, but skipping gracefully when not all required libraries are installed?) and have it built last (after all of the command-line tools have been built, so that a Johnny build failure wouldn't prevent access to any of the existing functionality of jumbo). This will also ensure that a user gets a compatible combination of versions of jumbo's command-line tools and Johnny (as long as we do maintain it).

To make it clear: the only merge I am proposing here is of the source trees and build procedures - have both projects combined into the same ./configure && make. Not any merge of the source code beyond the build scripts, etc. In fact, Johnny's source code should mostly remain in its own subdirectory in the merged tree (perhaps src/johnny or src/gui).

If there are no objections, then who's to work on it? Maybe @shinnok and/or @math07? :-)

**Comment by shinnok** [comment]
I think this has been discussed in the past and I approved of it then. Supplying both tools in a single package is still the best approach; for all other cases people have the option of installing core JtR and Johnny via the distribution specific packages (i.e. Kali). 

If @math07 or somebody else offers to pick this up I will mentor and assist, if not I can pick this up myself. The task should be simple and Johnny has only one extra requirement to check and build against in addition to what Jumbo already requires, which is the Qt libs, out of which only `qt5-core qt5-widgets qt5-qmake qt5-buildtools` are needed, thus a minimal extra requirement for end users.

**Comment by solardiz** [comment]
Relevant discussion appeared in https://github.com/magnumripper/JohnTheRipper/issues/4243#issuecomment-610135539 but should be brought to this issue.

@shinnok wrote:

> Not sure how many CLI packages "bundle" a GUI in the same source tree; I can see several packaging issues originating from this practice.

I think our closest match might be Bitcoin Core and its forks. Some of its forks dropped the GUI, though - Zcash did (and thus forks of Zcash too), perhaps to let them focus on their core product within limited budget and time.

Nmap+Zenmap appears to be somewhere in the middle: Zenmap is somewhat separate, but uses version numbers consistent with Nmap's (thus, is updated in sync) and is bundled along with Nmap in some binary packages.

> I'm having trouble remembering, what was the original driver behind bundling Johnny at the source level tree with Jumbo?

Here's my thinking:

1. With Johnny staying separate, it was kind of abandoned at times, then revived. With it being in the same tree, we'll hopefully maintain it and keep it in sync with John.

2. People sometimes make and distribute binary builds of John, but they never did it for Johnny so far (the only builds of Johnny are by developers of Johnny itself). Maybe the merging will prompt inclusion of Johnny in such builds.

3. GUI is for those who want ease of use. Having to download two separate components and get them to work together is against ease of use. The result is little overlap between those who would benefit from the GUI and those who would actually bother and succeed to get it installed and working.

Nmap's approach is also an option. With it, we can partially address 1 by keeping both repos nearby (perhaps under Openwall account on GitHub) and prominently linked from JtR homepage. We can address 2 by not neglecting to make builds for popular systems ourselves each time we make a release. We can address 3 by bundling the components together (and properly pre-configured) in some binary packages.

**Comment by shinnok** [comment]
Good find, I missed this issue. I'm dumping my mind here:

1. Good example with Zenmap, I actually worked on that many years ago. However, it being a Python app makes things much easier (no compilation) from the little I remember.
2. Regarding the second paragraph in my first comment [here](https://github.com/magnumripper/JohnTheRipper/issues/2673#issuecomment-322717655), it requires reconsideration. Qt is not a small dependency by any means, although modularity has greatly improved in recent years. Pulling Qt as dependency for Jumbo by default is certainly not something many distros might want to do. How many do distribute Jumbo?
3. If we don't leverage the dependency aspect of packaging, then we are left with the same problem of letting people know about the GUI and *that it can be enabled*. The visibility of the tool can be improved nonetheless via more prominent mentions in distribution files, openwall.org and moving to @openwall account (which I still own). A more involved hand in packaging and distribution might also help. See [johnny#121](https://github.com/shinnok/johnny/issues/121).
3. QMake is the default conf system for Qt (what Johnny also uses), however, by latest plans The Qt Company plans a switch to CMake. This might further complicate the integration since Jumbo doesn't use it either.

First steps I'm considering:

a) Test current state of Johnny with regards to latest Qt LTS, the build system, clear out the issue backlog and maybe develop a bit further.
b) The UI needs a quick refresh to bring it to the new decade.
c) Reconsider the bundling aspect after I've reacquainted myself with Johnny and its possibilities going forward.
d) If we don't decide to bundle the two at the source level, we still can distribute snaps with those two included right?
e) I don't think anyone actually tried to get Johnny in any repositories intently? Kali Linux picked it up on their own, but I'm sure we could get it in all major repositories where Jumbo is with little effort. This is arguably much better and future-proof than distributing binary packages ourselves. Windows is the obvious outsider here, where we do need to build binaries no matter the distribution channel.

**Comment by solardiz** [comment]
Of course, we must not introduce any new hard dependencies to jumbo. Any system that is able to build and run jumbo now should continue to be at least as capable of that after the possible merge of Johnny.

This should be similar to how we support e.g. OpenCL: if `configure` detects the required dependencies, it gets automatically enabled at build time (and the summary that `configure` prints at the end states so), otherwise it's left disabled. There are also `configure` options to explicitly enable OpenCL (and fail if dependencies aren't present?) or to explicitly disable it. (We should similarly turn OpenSSL into an optional dependency soon.)

The above implies that we must not bring the whole of jumbo onto whatever build system Qt needs - we need to use that other build system just for the Qt-using portion, and only when its build is enabled.

I think in terms of (lack of) build dependencies it's similar to Bitcoin Core, which can be built with or without its Qt GUI. [Their GUI is more deeply integrated, though: it's not a wrapper program, but is a full wallet. We could (re)consider something like that (much) later too, but given that JtR started as command-line only it's essential to keep the GUI as a wrapper around that (at least for now). In fact, some years before Johnny was born we had someone experiment privately with a deeply integrated GUI for JtR using wxWidgets, but this didn't lead to anything we could release in part (but not only) because JtR itself was a moving target whereas deep integration meant building upon a particular version of JtR (or needing to re-sync to new versions). So when I mention Bitcoin Core now, I only mean the approach to building it with or without the GUI, not the level of GUI integration with the core program.]

To your specific questions:

> Pulling Qt as dependency for Jumbo by default is certainly not something many distros might want to do. How many do distribute Jumbo?

I haven't been keeping track of how many distributions include Jumbo. Maybe I should, and maybe I should list that on JtR homepage, like we have it for several smaller Openwall projects. Anyhow, a problem is that some distributions will lag behind our releases. Debian is especially conservative.

For packaged builds (ours and/or distros') an option is to build several binary sub-packages out of the same source package. When built on a system that has Qt, both a CLI and a GUI sub-package would be created. Then on a system without Qt runtime it would nevertheless be possible to install and use the CLI sub-package.

> If we don't decide to bundle the two at the source level, we still can distribute snaps with those two included right?

I'm not familiar with snaps (I've never even used them), but I guess so. I expect this will be up to Claudio, as he's the one who brought snaps to our project and maintains those, and there doesn't appear to be anyone else currently willing to help with that. My gut feeling is that with Johnny staying separate, Claudio is less likely to take the effort of bringing it into the snaps, but I can't speak for him. It's also possible that even with Johnny merged, Claudio won't enable its build in the snaps (although I hope he would) - again, it's up to him and I can't speak for him, I merely appreciate whatever contributions he does make.

>  I don't think anyone actually tried to get Johnny in any repositories intently? Kali Linux picked it up on their own, but I'm sure we could get it in all major repositories where Jumbo is with little effort. This is arguably much better and future-proof than distributing binary packages ourselves.

Yes, as far as I'm aware none of us tried to get Johnny into any repositories. I'm not sure how little effort (or not) it would be. It's probably quite some effort. And it's arguable how future-proof that is: distro packages may be out of date, and at least Debian (and thus Ubuntu) has a policy preventing new upstream versions from getting into stable branches of the distros. This means that in those branches we won't even be able to get bugs fixed without working on back-ports of bug fixes, which is significant extra effort.

**Comment by magnumripper** [comment]
> 3. If we don't leverage the dependency aspect of packaging, then we are left with the same problem of letting people know about the GUI and _that it can be enabled_.

Come to think of it, we could (right now) output a hint at end of configure and/or make - perhaps we should have done so years ago.

**Comment by solardiz** [comment]
> output a hint at end of configure and/or make

Makes sense. It'd make even more sense to inform/remind of Johnny there once Johnny is hosted on the same GitHub account with jumbo, and on the main Openwall website - which is something I should make happen.
